### PR TITLE
[#10920] Add snapshot tests for admin search page

### PR DESCRIPTION
**Pull request by jianhandev**
Part of #10920

**Outline of solution:**
- Added snapshot tests for AdminSearchPageComponent
- Added ~~positive and~~ negative tests for `openCourseJoinEmail` and `openFeedbackSessionReminderEmail`

### CODE REVIEW

File: src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts [hunk 1]
```diff
@@ -555,4 +617,93 @@ describe('AdminSearchPageComponent', () => {
     expect(spyStatusMessageService).toBeCalled();
   });
 
+  it('should show error message if fail to send course join email', () => {
+    const studentResult: StudentAccountSearchResult = {
```
Comment (Derek-Hardy) [inline src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts:621]: Maybe we can add a positive test case in some way for this endpoint?

File: src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts [hunk 2]
```diff
@@ -555,4 +617,93 @@ describe('AdminSearchPageComponent', () => {
...
+        });
+
+    const sendButton: any = fixture.debugElement.nativeElement.querySelector('#send-course-join-button');
+    sendButton.click();
+
+    expect(spyStatusMessageService).toBeCalled();
+  });
+
+  it('should show error message if fail to send session reminder email', () => {
+    const studentResult: StudentAccountSearchResult = {
```
Comment (Derek-Hardy) [inline src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts:661]: Same for this endpoint.

File: src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts (hunk 1 above)
Comment (jianhandev) [reply src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts:621]: It seems quite tricky to test `window.location.href` without modifying non-test code, do you have any ideas/suggestions?

Comment (Derek-Hardy) [reply src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts:621]: Yup, I know. 

I think it will be sufficient to mock the response of `getCourseJoinEmail` and `getFeedbackSessionReminderEmail` method calls in `emailGenerationService`.

And then check service is called and values of `email.recipient`, `email.subject`, `email.content` are as expected.

File: src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts [hunk 3]
```diff
@@ -555,4 +618,179 @@ describe('AdminSearchPageComponent', () => {
...
+          recipient: 'Jacky Chan',
+          subject: 'Course join email',
+          content: 'Course join email content',
+        }),
+    );
+
+    const sendButton: any = fixture.debugElement.nativeElement.querySelector('#send-course-join-button');
+    sendButton.click();
+
+    expect(spyEmailGenerationService).toBeCalled();
```
Comment (jianhandev) [inline src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts:655]: Technically the email object is not returned or stored in the component afterwards. So I'm wondering if I am spying on the email generation service, is it possible to retrieve the email's recipient/subject/content?

**Comment by Derek-Hardy** [comment]
You seem to have missed the `id` in html for the `send-course-join-button`.

File: src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts (hunk 3 above)
Comment (Derek-Hardy) [reply src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts:655]: True, this indeed relies `window.location.href` which is not very straightforward to test...

Any suggestions? @ChooJeremy @t-cheepeng @madanalogy

**Comment by jianhandev** [comment]
> @jianhandev You seem to have missed the `id` in html file for the `send-course-join-button`. Which leads to some test failures.

Thanks for pointing that out!

File: src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts (hunk 3 above)
Comment (t-cheepeng) [reply src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts:655]: @jianhandev You could try to inject a `window` object into the test suite using a `InjectionToken` and test against that.

Some references:
https://angular.io/api/core/InjectionToken
https://jasminexie.github.io/injecting-window-in-an-angular-application/

Comment (jianhandev) [reply src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts:655]: Thanks for the suggestion! Will seriously look into that 👍

**Review by ChooJeremy — APPROVED** [review]
Tests LGTM, other than the issue mentioned on email generation service that uses `window.location.href`. Good coverage.

File: src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts [hunk 4]
```diff
@@ -36,10 +40,11 @@ describe('AdminSearchPageComponent', () => {
         FormsModule,
         HttpClientTestingModule,
         NgbTooltipModule,
+        BrowserAnimationsModule,
       ],
       providers: [AccountService, SearchService, StatusMessageService, NgbModal],
     })
-    .compileComponents();
+        .compileComponents();
```
Comment (madanalogy) [inline src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts:47]: Any particular reason for this extra indent?

File: src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts [hunk 5]
```diff
@@ -65,9 +128,9 @@ describe('AdminSearchPageComponent', () => {
     }));
 
     const spyStatusMessageService: any = spyOn(statusMessageService, 'showErrorToast').and.callFake(
-      (args: string): void => {
-        expect(args).toEqual('This is the error message');
-      });
+        (args: string): void => {
+          expect(args).toEqual('This is the error message');
+        });
```
Comment (madanalogy) [inline src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts:133]: Similarly here

File: src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts [hunk 6]
```diff
@@ -82,9 +145,9 @@ describe('AdminSearchPageComponent', () => {
     }));
 
     const spyStatusMessageService: any = spyOn(statusMessageService, 'showWarningToast').and.callFake(
-      (args: string): void => {
-        expect(args).toEqual('No results found.');
-      });
+        (args: string): void => {
+          expect(args).toEqual('No results found.');
+        });
```
Comment (madanalogy) [inline src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts:150]: And here

**Review by madanalogy — CHANGES_REQUESTED** [review]
I'm not usually particular about small unintended changes being made but there's really just too many occurrences in this PR.

File: src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts (hunk 4 above)
Comment (jianhandev) [reply src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts:47]: I think those were auto-indented by my IDE. It did managed to pass the lint test so I thought it was fine. Nevertheless, I will double check my IDE settings again for indentation. Thanks for pointing out!

**Comment by rrtheonlyone** [comment]
@jianhandev is this PR going to be ready for review?

**Comment by jianhandev** [comment]
> @jianhandev is this PR going to be ready for review?

@rrtheonlyone other than the issue mentioned on email generation service that uses `window.location.href` I think it is ready for a final review. I'm sort of stuck on that issue, would be nice if someone who's more experienced in front-end testing can pick it up and work on it.

File: src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts [hunk 7]
```diff
@@ -544,15 +604,189 @@ describe('AdminSearchPageComponent', () => {
...
+
+    const sendButton: any = fixture.debugElement.nativeElement.querySelector('#send-course-join-button');
+    sendButton.click();
+
+    expect(spyStatusMessageService).toBeCalled();
+  });
+
+  it('should generate email when send session reminder email button clicked', () => {
+    const studentResult: StudentAccountSearchResult = {
+      name: 'Jack Chan',
```
Comment (madanalogy) [inline src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts:696]: Could we abstract the `StudentAccountSearchResult` objects in this file into a factory function of sorts, and you can populate the necessary fields as parameters. Seems to me like not many fields need to change between test cases. For example, 

```
const initStudentSearchResult = (name, email) => {
  return {
    // populate data

File: src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts [hunk 8]
```diff
@@ -471,8 +531,8 @@ describe('AdminSearchPageComponent', () => {
       newRegistrationKey: 'newKey',
     }));
 
-    const spyStatusMessageService: any = spyOn(statusMessageService, 'showSuccessToast').and.callFake(
-        (args: string): void => {
+    const spyStatusMessageService: any = spyOn(statusMessageService, 'showSuccessToast')
+        .and.callFake((args: string) => {
```
Comment (madanalogy) [inline src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts:535]: Still seems like a bunch of unnecessary changes in this file

**Review by madanalogy — CHANGES_REQUESTED** [review]

File: src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts (hunk 5 above)
Comment (jianhandev) [reply src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts]: @madanalogy Thanks for the suggestion, made the changes 👍

File: src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts (hunk 7 above)
Comment (jianhandev) [reply src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts]: @madanalogy Thanks for the suggestion, made the changes 👍

File: src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts (hunk 8 above)
Comment (jianhandev) [reply src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts:504]: The re-indentation was done because it exceeds the line limit, I double-checked my indentation settings in my IDE and seems to be correct also (I'm following what is specified in https://github.com/TEAMMATES/teammates/blob/master/docs/ide-setup.md).

**Review by madanalogy — APPROVED** [review]

**Review by rrtheonlyone — APPROVED** [review]
